exonum-java-binding/common/src/main/java/com/exonum/binding/common/proofs/map/UncheckedFlatMapProof.java

```diff
@@ -249,42 +257,114 @@ private CheckedMapProof checkProof() {
-    return HashCode.fromBytes(new byte[Hashing.DEFAULT_HASH_SIZE_BYTES]);
+  private static HashCode getEmptyProofIndexHash() {
+    HashCode merkleRoot = HashCode.fromBytes(new byte[Hashing.DEFAULT_HASH_SIZE_BYTES]);
+    return getIndexHash(merkleRoot);
   }
 
-  private static HashCode getSingleEntryRootHash(MapProofEntry entry) {
-    return getSingleEntryRootHash(entry.getDbKey(), entry.getHash());
+  private static HashCode getSingleEntryProofIndexHash(MapProofEntry proofEntry) {
+    HashCode merkleRoot = getSingleEntryMerkleRoot(proofEntry.getDbKey(),
+        proofEntry.getHash());
+    return getIndexHash(merkleRoot);
   }
 
-  private static HashCode getSingleEntryRootHash(MapEntry<ByteString, ByteString> entry) {
-    DbKey dbKey = DbKey.newLeafKey(entry.getKey());
-    HashCode valueHash = HASH_FUNCTION.hashByteString(entry.getValue());
-    return getSingleEntryRootHash(dbKey, valueHash);
+  private static HashCode getSingleEntryProofIndexHash(MapEntry<ByteString, ByteString> mapEntry) {
+    DbKey dbKey = DbKey.newLeafKey(mapEntry.getKey());
+    HashCode valueHash = getLeafEntryHash(mapEntry.getValue());
+    HashCode merkleRoot = getSingleEntryMerkleRoot(dbKey, valueHash);
+    return getIndexHash(merkleRoot);
   }
 
-  private static HashCode getSingleEntryRootHash(DbKey key, HashCode valueHash) {
+  private static HashCode getIndexHash(HashCode merkleRoot) {
+    return HASH_FUNCTION.newHasher()
+        .putByte(MAP_ROOT_PREFIX)
+        .putObject(merkleRoot, hashCodeFunnel())
+        .hash();
+  }
+
+  private static HashCode getSingleEntryMerkleRoot(DbKey key, HashCode valueHash) {
     assert key.getNodeType() == Type.LEAF;
     return HASH_FUNCTION.newHasher()
+        .putByte(MAP_NODE_PREFIX)
         .putObject(key, dbKeyFunnel())
         .putObject(valueHash, hashCodeFunnel())
         .hash();
   }
 
-  private static HashCode getMapEntryHash(MapEntry<ByteString, ByteString> entry) {
-    return HASH_FUNCTION.hashByteString(entry.getValue());
+  private static HashCode getLeafEntryHash(ByteString entryValue) {
+    return HASH_FUNCTION.newHasher()
+        .putByte(BLOB_PREFIX)
+        .putBytes(entryValue.toByteArray())
+        .hash();
   }
 
   private static HashCode computeBranchHash(MapProofEntry leftChild, MapProofEntry rightChild) {
+    byte[] leftChildKey = compress(leftChild.getDbKey());
+    byte[] rightChildKey = compress(rightChild.getDbKey());
     return HASH_FUNCTION
         .newHasher()
+        .putByte(MAP_NODE_PREFIX)
         .putObject(leftChild.getHash(), hashCodeFunnel())
         .putObject(rightChild.getHash(), hashCodeFunnel())
-        .putObject(leftChild.getDbKey(), dbKeyFunnel())
-        .putObject(rightChild.getDbKey(), dbKeyFunnel())
+        .putBytes(leftChildKey)
+        .putBytes(rightChildKey)
         .hash();
   }
 
+  /**
+   * Returns a LEB128 compressed binary representation of the given db key.
+   */
+  public static byte[] compress(DbKey dbKey) {
```

Comment:
Why isn't it in a funnel so that it is more easily testable?

On top of that, it will provide easier API to write the things (putByte/putBytes, similar to ByteBuffer, instead of numerous array copies and keeping track of the number of written bytes manually).

Please also test this.